### PR TITLE
Chat functionality

### PR DESCRIPTION
**Pull request by davidmohar**

### CODE REVIEW

File: world-server/src/world/player/Player.ts [hunk 1]
```diff
@@ -12,6 +12,7 @@ export class Player extends Entity {
     public weapon: Weapon;
     public score: number = 0;
     public direction: number = 0;
+    public initialposition: Vector2;
```
Comment (jakakonda) [inline world-server/src/world/player/Player.ts]: initialPosition...
How is it related to chat? Why do we actually need to store it?

File: world-server/src/world/World.ts [hunk 2]
```diff
@@ -10,9 +10,12 @@ import { Router } from 'world-core';
 import random = require('random');
 import seedrandom = require('seedrandom');
 import SimplexNoise = require('simplex-noise');
+import { ChatHandler } from "./chat/ChatHandlers";
+import { router } from "websocket";
```
Comment (jakakonda) [inline world-server/src/world/World.ts]: Where is this used?

File: world-server/src/world/chat/ChatHandlers.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import { Messages, Router, Logger } from 'world-core';
+import { BroadcastMessage } from '../../network';
+import { ChatMessage } from './ChatMessages';
+import { Player } from '../player';
+
+export class ChatHandler implements Messages.MsgHandler {
+    private players: Player[];
+
+    constructor(players: Player[]) {
```
Comment (jakakonda) [inline world-server/src/world/chat/ChatHandlers.ts]: This parameter is copy & pasted and not actually used anywhere. I believe it should be removed.

File: world-server/src/world/chat/ChatHandlers.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+        return true;
+    }
+
+    public handle(msg: Messages.Message): void {
+        let message = msg as ChatMessage;
+        switch (message.type) {
+            case 'chat.say':
+                let chatMessage = message as ChatMessage;
+                chatMessage.id = msg.clientId;
```
Comment (jakakonda) [inline world-server/src/world/chat/ChatHandlers.ts:44]: Shouldn't we assign a unique id to chat messages?

File: world-server/src/world/player/Player.ts [hunk 5]
```diff
@@ -60,10 +63,16 @@ export class Player extends Entity {
 
     respawn(){
         // TODO: respawn logic
+        this.health = this.maxHealth;
```
Comment (jakakonda) [inline world-server/src/world/player/Player.ts]: Should be in a separate pull request...

**Review by jakakonda — CHANGES_REQUESTED** [review]
Missing fronted. Also see comments.

File: world-server/src/world/player/Player.ts (hunk 1 above)
Comment (filiptrplan) [reply world-server/src/world/player/Player.ts]: The guys working on that seemed to forgot to checkout to a separate branch when working on the respawn functionality. Removed.

File: world-server/src/world/World.ts (hunk 2 above)
Comment (filiptrplan) [reply world-server/src/world/World.ts]: `router` is not used, removed. The chat handler is registered in the `World constructor` function.

File: world-server/src/world/player/Player.ts (hunk 5 above)
Comment (filiptrplan) [reply world-server/src/world/player/Player.ts]: Same thing as with the initial position. It seems the guys got confused, removed.

File: world-server/src/world/chat/ChatHandlers.ts (hunk 4 above)
Comment (filiptrplan) [reply world-server/src/world/chat/ChatHandlers.ts:44]: Added

**Comment by filiptrplan** [comment]
Don't merge yet.
The keys WASD don't type any characters in the chat box due to some weird input hooking done by Phaser. @duhecx should know more about how to solve this.

File: world-server/src/world/chat/ChatHandlers.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    public getTypes(): string[] {
+        return ['chat.say'];
+    }
+
+    public validate(msg: Messages.Message): boolean {
+        let message = msg as ChatMessage;
+        if (message == null || message == undefined) {
+            Logger.warn(`Received player message does not extend ChatMessage`);
```
Comment (jakakonda) [inline world-server/src/world/chat/ChatHandlers.ts]: Technically incorrect comment. Casts in Typescript are only for developers. At the end, JS is executed and no runtime type checks are present.

File: world-server/src/world/chat/ChatHandlers.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+            return false;
+        }
+        if (message.text == null || message.text == undefined) {
```
Comment (jakakonda) [inline world-server/src/world/chat/ChatHandlers.ts]: if (!message.text) solves this and the next if.

File: world-client/src/features/chat.js [hunk 8]
```diff
@@ -13,6 +13,46 @@ export default class Chat extends Feature {
...
+      if (e.key == "Enter"){
+        wop.socket.send({
+          type: 'chat.say',
+          id: wop.me.id,
+          name : wop.me.name,
+          text : inputMessage.value,
+        });        
+        inputMessage.value = "";
+        inputMessage.blur();
+        console.log("blur");
```
Comment (aJuvan) [inline world-client/src/features/chat.js]: Remove debug logs

File: world-client/src/features/chat.js [hunk 9]
```diff
@@ -23,14 +63,16 @@ export default class Chat extends Feature {
...
     // On server message received logic here
     switch (message.type) {
-      case "player.say":
+      case "chat.say":
         // Player says something in the chat
-        
+        // <li>Enej: Hej</li>
+        var messages = window.document.getElementById("messages");
+        var li = window.document.createElement("li");
+        li.innerHTML = "<b>" + message.name + ":</b> " + message.text;
```
Comment (jakakonda) [inline world-client/src/features/chat.js]: XSS vulnerability

File: world-server/src/world/chat/ChatHandlers.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+            return false;
+        }
+
+        if (message.text.length == 0) {
```
Comment (jakakonda) [inline world-server/src/world/chat/ChatHandlers.ts]: We should probably add max length constraint.

**Review by jakakonda — CHANGES_REQUESTED** [review]

File: world-server/src/world/chat/ChatHandlers.ts (hunk 6 above)
Comment (filiptrplan) [reply world-server/src/world/chat/ChatHandlers.ts]: I didn't write this but I guess that this could be triggered if a client sends a `chat.say` message but uses message contents for another message type. Maybe we could check this another way so it actually works? 

Could be a helper function that is written so that we can use it on all validate functions?

File: world-server/src/world/chat/ChatHandlers.ts (hunk 7 above)
Comment (filiptrplan) [reply world-server/src/world/chat/ChatHandlers.ts]: I merged the two into one if statement

File: world-server/src/world/chat/ChatHandlers.ts (hunk 6 above)
Comment (filiptrplan) [reply world-server/src/world/chat/ChatHandlers.ts]: Asked @davidmohar about this, says it's impossible, removing the logger warning.

**Review by aJuvan — APPROVED** [review]

**Review by jakakonda — APPROVED** [review]
